### ISSUE TITLE
llvm-bolt –reg-reassin core dump when the input binary(mysqld) is PGO optimized

### ISSUE THREAD

**Issue by llongint** [llvm/llvm-project#59730]
It's been described [here](https://discourse.llvm.org/t/llvm-bolt-reg-reassin-core-dump-when-the-input-binary-mysqld-is-pgo-optimized/67380) before.

[test.s.txt](https://github.com/llvm/llvm-project/files/10312663/test.s.txt)
[test.fdata.txt](https://github.com/llvm/llvm-project/files/10312664/test.fdata.txt)

Run the following command to reproduce the problem:
```bash
mv test.s.txt test.s
gcc -O0 test.s -Wl,-q -o a.out
./a.out
llvm-bolt ./a.out -o ./a.opt -data=./test.fdata.txt --reg-reassign --print-regreassign -debug-only=regreassign
./a.opt
```
./a.opt will fall into an infinite loop.
![image](https://user-images.githubusercontent.com/28920187/209771991-cd0695f7-102e-465d-9090-3a2c9f7f0de6.png)

![image](https://user-images.githubusercontent.com/28920187/209772272-bad064a6-25e6-4a10-b5cb-cd5c82fb7e91.png)

**Comment by llongint** [comment]
@aaupov

**Comment by aaupov** [comment]
@llongint - sorry about the delay. I will get back to this issue next week. In the meantime you can disable this optimization as layout optimizations generally provide much higher value.

**Comment by llongint** [comment]
Thanks.  That's okay, but it looks like rafaelauler is already looking at this issue.  See [here](https://discourse.llvm.org/t/llvm-bolt-reg-reassin-core-dump-when-the-input-binary-mysqld-is-pgo-optimized/67380/7?u=hezuoqiang)

**Comment by aaupov** [comment]
@llongint – the problem is that BOLT doesn't perform cross-function reg-reassign. It would be better to disable function splitting in the compiler so that BOLT performs it using the accurate profile. That may in turn enable reg-reassign for this particular case.

**Comment by llongint** [comment]
That should work, but I haven't found an GCC option to turn off.

Also, I noticed that BOLT gives this hint, if this issue cannot be solved, maybe we can exit this optimization when this is detected?
```log
BOLT-WARNING: debug info will be stripped from the binary. Use -update-debug-sections to keep it.
```

**Comment by aaupov** [comment]
The flag to disable GCC function splitting is `-fno-reorder-blocks-and-partition`.
Split function support is by best effort, and reg-reassign is an experimental optimization. The two weren't extensively tested together.

**Comment by llongint** [comment]
Can we simply fix it like this [D141931](https://reviews.llvm.org/D141931)
